### DISCUSSION TITLE
Protothread and Simple UDP

### DISCUSSION THREAD

**Discussion by rahav** [contiki-ng/contiki-ng#2171] [Q&A]
The docs state that a protothread will never get interrupted unless a hardware interrupt occurred. I'm using Simple UDP for communication. Is the arrival of a new message considered a hardware interrupt? That us, if a `process` (protothread) is running when a message arrives, will the `process` be interrupted and the `rx_callback` function called?

If the `process` is interrupted, where can I expect control to be after we return from the `rx_callback` function, back at the interrupted `process`, or is the scheduler allowed to invoke any other `process` first?

**Comment by alexrayne** [comment]
contiki protothreads - are cooperative multi-tasking, with no preemption - your code full controls on switching your tasks. Interupts are preemts any ordinar code, ans usual. So you have to inform your tasks about interupt events, and it on your own response. Contiki gives you only 1 method for this - `process_poll`.  Your have to desin protothread code to recognise that events. look for examples, and drivers code - to see the best practises.

**Reply by rahav** [reply]
Thank you Alex for your reply. I'm not sure I understand your answer. I will try to ask it differently.

1. Is the arrival of a new message going to interrupt a running protothread?
